vortex-scalar/src/primitive.rs

```diff
@@ -270,3 +272,47 @@ impl From<usize> for Scalar {
         Scalar::primitive(value as u64, Nullability::NonNullable)
     }
 }
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum NumericOperator {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    // Missing from arrow-rs:
+    // Min,
+    // Max,
+    // Sqrt,
+    // Pow,
+    // In IEEE-754 operations involving NaN are non-commutative wrt NaN payload/quietness
+    // RAdd,
+    // RSub,
+    // RMul,
+    // RDiv,
+}
+
+impl PrimitiveScalar<'_> {
+    pub fn numeric_operator(
+        self,
+        other: PrimitiveScalar<'_>,
+        op: NumericOperator,
+    ) -> VortexResult<Scalar> {
+        if !self.dtype().eq_ignore_nullability(other.dtype()) {
+            vortex_bail!("types must match: {} {}", self.dtype(), other.dtype());
+        }
+
+        Ok(match_each_native_ptype!(self.ptype(), |$P| {
+            let lhs = self.typed_value::<$P>();
+            let rhs = other.typed_value::<$P>();
+            let nullability = self.dtype().nullability();
+            lhs.zip(rhs).map(|(lhs, rhs)| {
```

Comment:
Zipping optionals is quite hard to parse... I guess it means if either side is None then the result is an empty iterator?